loader/terminator.c

```diff
@@ -626,23 +560,31 @@ VKAPI_ATTR VkResult VKAPI_CALL terminator_GetPhysicalDeviceToolProperties(VkPhys
                                                                           VkPhysicalDeviceToolProperties *pToolProperties) {
     struct loader_physical_device_term *phys_dev_term = (struct loader_physical_device_term *)physicalDevice;
     struct loader_icd_term *icd_term = phys_dev_term->this_icd_term;
+    VkResult res = VK_SUCCESS;
 
-    if (NULL == icd_term->dispatch.GetPhysicalDeviceToolProperties) {
-        loader_log(icd_term->this_instance, VULKAN_LOADER_ERROR_BIT, 0,
-                   "terminator_GetPhysicalDeviceToolProperties: The ICD's vkGetPhysicalDeviceToolProperties was NULL yet "
-                   "the physical device supports Vulkan API Version 1.3.");
-    } else {
-        VkPhysicalDeviceProperties properties;
-        if (icd_term->dispatch.GetPhysicalDeviceProperties) {
-            icd_term->dispatch.GetPhysicalDeviceProperties(phys_dev_term->phys_dev, &properties);
+    if (pToolCount != NULL && pToolProperties == NULL) {
+        *pToolCount = 0;
+    }
 
-            if (VK_API_VERSION_MINOR(properties.apiVersion) >= 3) {
-                return icd_term->dispatch.GetPhysicalDeviceToolProperties(phys_dev_term->phys_dev, pToolCount, pToolProperties);
-            }
+    // If this is Vulkan 1.3 or newer, then it may support the API (but then again, it may not).
+    // Because of this, check for NULL before trying to call down.
+    if (VK_API_VERSION_MINOR(phys_dev_term->properties.apiVersion) >= 3) {
+        if (NULL == icd_term->dispatch.GetPhysicalDeviceToolProperties) {
+            loader_log(icd_term->this_instance, VULKAN_LOADER_ERROR_BIT | VULKAN_LOADER_DRIVER_BIT, 0,
+                       "terminator_GetPhysicalDeviceToolProperties: Driver \"%s\" device \"%s\" "
+                       "vkGetPhysicalDeviceToolProperties was NULL yet device claims supports for Vulkan 1.3",
+                       icd_term->scanned_icd->lib_name, phys_dev_term->properties.deviceName);
+        } else {
+            res = icd_term->dispatch.GetPhysicalDeviceToolProperties(phys_dev_term->phys_dev, pToolCount, pToolProperties);
+            goto out;
         }
     }
 
-    // In the case the driver didn't support 1.3, make sure that the first layer doesn't find the count uninitialized
-    *pToolCount = 0;
-    return VK_SUCCESS;
+    // Try using the extension version if it's available
+    if (NULL != icd_term->dispatch.GetPhysicalDeviceToolPropertiesEXT) {
+        res = icd_term->dispatch.GetPhysicalDeviceToolPropertiesEXT(phys_dev_term->phys_dev, pToolCount, pToolProperties);
+    }
```

Comment:
This is insufficient for use with the extension `vkGetPhysicalDeviceToolPropertiesEXT` unless I misundersood something.

The reason is that we cannot rely on drivers non-NULL `vkGetPhysicalDeviceToolPropertiesEXT` meaning its safe to call. Because this function used to not call into drivers, it is on us to verify that we don't call drivers which might not actually support it. The old extension code would search the list of extensions available on a physical device first. This works around MESA reporting `vkGetPhysicalDeviceToolPropertiesEXT` even if there are no drivers which support the function.